src/System.Memory/src/System/Text/EncodingExtensions.cs

```diff
@@ -0,0 +1,655 @@
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace System.Text
+{
+    public static class EncodingExtensions
+    {
+        /// <summary>
+        /// The maximum number of input elements after which we'll begin to chunk the input.
+        /// </summary>
+        /// <remarks>
+        /// The reason for this chunking is that the existing Encoding / Encoder / Decoder APIs
+        /// like GetByteCount / GetCharCount will throw if an integer overflow occurs. Since
+        /// we may be working with large inputs in these extension methods, we don't want to
+        /// risk running into this issue. While it's technically possible even for 1 million
+        /// input elements to result in an overflow condition, such a scenario is unrealistic,
+        /// so we won't worry about it.
+        /// </remarks>
+        private const int MAX_INPUT_ELEMENTS_PER_ITERATION = 1 * 1024 * 1024;
+
+        /// <summary>
+        /// Encodes the specified <see cref="ReadOnlySpan{Char}"/> to <see langword="byte"/>s using the specified <see cref="Encoding"/>
+        /// and writes the result to <paramref name="writer"/>.
+        /// </summary>
+        /// <param name="encoding">The <see cref="Encoding"/> which represents how the data in <paramref name="chars"/> should be encoded.</param>
+        /// <param name="chars">The <see cref="ReadOnlySequence{Char}"/> to encode to <see langword="byte"/>s.</param>
+        /// <param name="writer">The buffer to which the encoded bytes will be written.</param>
+        /// <exception cref="EncoderFallbackException">Thrown if <paramref name="chars"/> contains data that cannot be encoded and <paramref name="encoding"/> is configured
+        /// to throw an exception when such data is seen.</exception>
+        public static long GetBytes(this Encoding encoding, ReadOnlySpan<char> chars, IBufferWriter<byte> writer)
+        {
+            if (encoding is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.encoding);
+            }
+
+            if (writer is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.writer);
+            }
+
+            if (chars.Length <= MAX_INPUT_ELEMENTS_PER_ITERATION)
+            {
+                // The input span is small enough where we can one-shot this.
+
+                int byteCount = encoding!.GetByteCount(chars);
+                Span<byte> scratchBuffer = writer!.GetSpan(byteCount);
+                Debug.Assert(scratchBuffer.Length >= byteCount, "Scratch buffer was shorter than expected.");
+
+                int actualBytesWritten = encoding.GetBytes(chars, scratchBuffer);
+                Debug.Assert(actualBytesWritten == byteCount, "Wrote unexpected number of bytes. Was buffer mutated unexpectedly?");
```

Comment:
Ditto

Reply:
This assert I think still adds value. It's checking for consistency between the two "byte count" values. If there's an inconsistency it means either that there's an error either in the `Encoding` instance provided or that the input data was unexpectedly mutated.

It's not fatal to this method if this occurs; the method will continue to make forward progress. But this is a sign that something unexpected happened and we probably generated malformed output data.

Reply:
From my perspective, I should never be able to trigger an assert based on arguments I supply to a public method, since the assert should only be for something that is categorically true. Yet here I can construct an Encoding that will cause this assert to fire. How do I write a test case for this if doing so will trigger the assert? If something unexpected happened and we generated malformed output data and we care about that, shouldn't that result in an exception rather than being silent in release builds?

Reply:
> If something unexpected happened and we generated malformed output data and we care about that, shouldn't that result in an exception rather than being silent in release builds?

This method doesn't care about that. I don't think the framework commonly performs such checks, since at a certain point it's infeasible to perform correctness checks and we just have to assume that the APIs we're calling (even if they're virtual methods) are properly implemented. But point taken, I can jettison those asserts.

Reply:
thanks :)